src/Microsoft.AspNetCore.Sockets/Internal/ConnectionState.cs

```diff
@@ -26,23 +34,54 @@ public ConnectionState(Connection connection, IChannelConnection<Message> applic
 
         public async Task DisposeAsync()
         {
-            // If the application task is faulted, propagate the error to the transport
-            if (ApplicationTask.IsFaulted)
+            Task applicationTask = TaskCache.CompletedTask;
+            Task transportTask = TaskCache.CompletedTask;
+
+            try
             {
-                Connection.Transport.Output.TryComplete(ApplicationTask.Exception.InnerException);
-            }
+                await Lock.WaitAsync();
+
+                if (Status == ConnectionStatus.Disposed)
+                {
+                    return;
+                }
+
+                Status = ConnectionStatus.Disposed;
+
+                RequestId = null;
 
-            // If the transport task is faulted, propagate the error to the application
-            if (TransportTask.IsFaulted)
+                // If the application task is faulted, propagate the error to the transport
+                if (ApplicationTask.IsFaulted)
+                {
+                    Connection.Transport.Output.TryComplete(ApplicationTask.Exception.InnerException);
+                }
+
+                // If the transport task is faulted, propagate the error to the application
+                if (TransportTask.IsFaulted)
+                {
+                    Application.Output.TryComplete(TransportTask.Exception.InnerException);
+                }
+
+                Connection.Dispose();
+                Application.Dispose();
+
+                applicationTask = ApplicationTask;
+                transportTask = TransportTask;
+            }
+            finally
             {
-                Application.Output.TryComplete(TransportTask.Exception.InnerException);
+                Lock.Release();
             }
 
-            Connection.Dispose();
-            Application.Dispose();
-
             // REVIEW: Add a timeout so we don't wait forever
```

Comment:
Just add it now :). All it takes is a Task.Delay (and maybe a check to fire TimeoutException if the delay task is the one that fires first).